load data files: update hints in error msgs and ignore the download by default

pirat89 opened the pull request:
As the leapp upgrade data files are nowadays part of the install rpm, there is no need to download them anymore. Also, we do not have any plans to provide the updated data files anywhere out of the rpm.

For that reason, update all texts and related error messages so people do not try to visit obsoleted article, trying to apply invalid (obsoleted) data files anymore.

I kept the original fetch functionality in case someone would like to use it for additional custom data files, but the load_data_asset function invoke the fetch without allowed downloads, so the fetch fails with the proper error msg if files are not stored locally.

@oamg/developers  postponed for the May release. Mostly covered by pr #1121 (changed err msgs). Another changes in this one PR are expected later after the release.

jira: RHEL-18298

MichalHe commented on repos/system_upgrade/common/libraries/fetch.py:
```diff
@@ -57,7 +71,8 @@ def read_or_fetch(filename,
                   encoding='utf-8',
-                  data_stream=None):
+                  data_stream=None,
+                  allow_download=True):
```
PR title says 'Ignore download by default' so this should be False, or?

pirat89 replied:
@MichalHe  good point. It depends. by default I am taking it as the behaviour when running `leapp (pre)upgrade`, however, from the function perspective, it made sense to me to keep true to protect original behaviour, but change the invocation of this function. From the perspective of our repositories, it's practically same, but it could have effect on 3rd-party content. wdyt in this case? Do we want to go here with the default False or True? I incline more to the second way due to this, but switching it to `False` as a default is also reasonable, from the POV that people should be aware of the change.

I will think about it yet, as this is small change, I will keep the decision about that as the last thing, for now, I will prepare the rest of the code and will see what we choose in the end. Anyway, nice catch during the review :-) I like that you think about it.

pirat89 replied:
@oamg/developers ^ headsup, in case you want to add your opinion :)

btw, it will be a short timer - this is going to be merged in couple of days (definitely this week).

pirat89 replied:
ah. just realized that part of that has been covered by https://github.com/oamg/leapp-repository/pull/1121 - which means the `allow_download=True` is already present publicly. so I would stay keeping it as this and change the invocation only.

pirat89 replied:
My apologize for the noise :pray: :brain:

pirat89 commented:
Reviving the work on this.

pirat89 commented:
Outputs from manual testing (note: required data stream is now (in time of testing) 2.x)
* set repomap.json data stream to 1.0, pes-events.json to 4.0, set empty data stream in "DDDD".json file

```
Risk Factor: high (inhibitor)
Title: Incompatible Leapp data assets are present
Summary: The currently installed Leapp consumes data stream 2.0, but the following assets provide different streams:
The following assets are outdated
 - The asset repomap.json provides the data stream 1.0
The following assets are intended for a newer leapp
 - The asset pes-events.json provides the data stream 4.0
Remediation: [hint] All official assets (data files) are nowadays part of the installed rpms. This issue is usually encountered when the data files are incorrectly customized, replaced, or removed.  In case you want to recover original files, remove them (if still exist) and reinstall following rpms: leapp-upgrade-el7toel8.
The listed assets (data files) are usually inside the /etc/leapp/files/ directory.
Key: a064e45f1d80af13c7c41f28cf1cf2a2e0c50359
----------------------------------------
Risk Factor: high (inhibitor)
Title: Detected malformed Leapp data assets
Summary: The following assets are malformed:
 - The asset file device_driver_deprecation_data.json contains no values in its "provided_data_streams" field, or the field does not contain a list
Remediation: [hint] All official assets (data files) are nowadays part of the installed rpms. This issue is usually encountered when the data files are incorrectly customized, replaced, or removed.  In case you want to recover original files, remove them (if still exist) and reinstall following rpms: leapp-upgrade-el7toel8.
The listed assets (data files) are usually inside the /etc/leapp/files/ directory.
Key: c5f9381e4725700584c5b45cb3cfa04d84e8be63
----------------------------------------
```

tests with various other malformations & removal:
```
# echo "" > device_driver_deprecation_data.json  # invalid json format 
# rm pes-events.json  # missing file
#  # remove provided_data_streams from repomap.json

## ..... in report ....:

Risk Factor: high (error)
Title: The Device driver deprecation data file (at device_driver_deprecation_data.json) does not contain a valid JSON object.
Summary: {"hint": "All official data files are nowadays part of the installed rpms. That is the only official resource of actual official data files for in-place upgrades. This issue is usually encountered when the data files are incorrectly customized, replaced, or removed (e.g. by custom scripts). In case you want to recover the original /etc/leapp/files/device_driver_deprecation_data.json file, remove the current one (if still exists) and reinstall following packages: leapp-upgrade-el7toel8."}
Key: 6ab1ae455d68a7bce41d2f6f4017712262b32516
----------------------------------------
Risk Factor: high (error)
Title: Data file /etc/leapp/files/pes-events.json is missing or invalid.
Summary: {"details": "File /etc/leapp/files/pes-events.json does not exist.", "hint": "All official data files are nowadays part of the installed rpms. That is the only official resource of actual official data files for in-place upgrades. This issue is usually encountered when the data files are incorrectly customized, replaced, or removed (e.g. by custom scripts). In case you want to recover the original /etc/leapp/files/pes-events.json file, remove the current one (if still exists) and reinstall following packages: leapp-upgrade-el7toel8."}
Key: c305148ff23ccf107bb5fe7acde505db4dcff88a
----------------------------------------

```

btw, the error representation in txt format seems little bit weird to me as I would expect from the past something like:
```
Risk:
Title:
Summary:
  Hint:
  Details:
  ...
```
maybe I remember it wrong. I will check that yet. Most likely this is thing of the framework and it is not related to changes in this PR at all.

pirat89 commented:
@fernflower  rebased to cover changes in testing: #1162

abadger reviewed (approved):
Looks good!  I made some suggestions but none of them are blockers.

Approved.

abadger commented on repos/system_upgrade/common/actors/checkconsumedassets/libraries/check_consumed_assets.py:
```diff
@@ -69,13 +86,16 @@ def report_incompatible_assets(assets):
         reporting.Severity(reporting.Severity.HIGH),
-        reporting.Groups([reporting.Groups.INHIBITOR, reporting.Groups.REPOSITORY]),
+        reporting.Remediation(hint=_get_hint()),
+        reporting.Groups([reporting.Groups.INHIBITOR, reporting.Groups.SANITY]),
     ]
 
     report_parts += make_report_entries_with_unique_urls(docs_url_to_title_map)
```
I went too quick and messed that up.  This is what I meant:

```suggestion
        reporting.Groups([reporting.Groups.INHIBITOR, reporting.Groups.SANITY]),
        make_report_entries_with_unique_urls(docs_url_to_title_map),
    ]

```

pirat89 replied:
discussed via slack that this would lead to a unexpected nested list inside `report_parts` of having a flat list. Both agreed to keep it rather unchanged.

pirat89 commented on repos/system_upgrade/common/actors/checkconsumedassets/libraries/check_consumed_assets.py:
```diff
@@ -101,13 +121,16 @@ def report_malformed_assets(malformed_assets):
+        reporting.Remediation(hint=_get_hint()),
         reporting.Severity(reporting.Severity.HIGH),
-        reporting.Groups([reporting.Groups.INHIBITOR, reporting.Groups.REPOSITORY]),
+        reporting.Groups([reporting.Groups.INHIBITOR, reporting.Groups.SANITY]),
     ]
 
     report_parts += make_report_entries_with_unique_urls(docs_url_to_title_map)
```
discussed via slack that this would lead to a unexpected nested list inside `report_parts` of having a flat list. Both agreed to keep it rather unchanged.

abadger commented on repos/system_upgrade/common/actors/peseventsscanner/libraries/pes_event_parsing.py:
```diff
@@ -83,22 +86,27 @@ def get_pes_events(pes_json_directory, pes_json_filename):
-            .format(rpmname)
+        )
+        hint = (
+            ' In case you want to recover the original {lp} file, remove it (if still exists)'
+            ' and reinstall following rpms: {rpms}.'
```
This suggestion did not seem to survive the rebase?  (Or maybe it's a github syncing issue and I'm just not seeing it in the webui).

pirat89 replied:
sorry for that. My bad. It's github webUI (syncing ?) issue as I see the butten is shadowed, saying the change has been applied. but the change is not in the branch. I should check that better. updating manually

abadger reviewed (approved):
Approved!
